R/eval-walk.R

```diff
@@ -286,12 +286,23 @@ call_kind <- function(expr, error_call) {
   fn <- as_string(head)
 
   if (fn %in% c("$", "[[") && identical(node_cadr(expr), quote(.data))) {
-    lifecycle::deprecate_warn("1.2.0",
-      what = I("Use of .data in tidyselect expressions"),
-      with = I("`any_of()` or `all_of()`")
-    )
-
     validate_dot_data(expr, error_call)
+
+    what <- I("Use of .data in tidyselect expressions")
+    if (fn == "$") {
+      var <- as.character(expr[[3]])
+      str <- encodeString(var, quote = '"')
+
+      lifecycle::deprecate_warn("1.2.0", what,
+        details = glue("Please use `{str}` instead of `data${var}`")
+      )
+    } else if (fn == "[[") {
+      var <- expr_deparse(expr[[3]])
+      lifecycle::deprecate_soft("1.2.0", what,
+        details = glue("Please use `all_of({var}) (or `any_of({var}) instead of `.data[[{var}]]`")
```

Comment:
```suggestion
        details = glue("Please use `all_of({var})` (or `any_of({var})` instead of `.data[[{var}]]`.")
```

```suggestion
        details = cli::format_inline("Please use {.code all_of({var})} (or {.code any_of({var})} instead of {.code .data[[{var}]]}.")
```